File: TimberCruisingApp/Geo/KMLImporter.swift
// Spec §8 Geo/KMLImporter. REQ-PRJ-002: import stratum boundaries from a
// Google-Earth-style KML document. Only `Placemark` elements carrying a
// `Polygon` (optionally nested under `MultiGeometry`) are extracted; other
// geometry types (Point/LineString/Model/etc.) are skipped.
//
// KML coordinates are expressed as `lon,lat[,alt]` whitespace-separated
// tuples. We reuse `ImportedPolygon` from `GeoJSONImporter` so downstream
// CruiseDesign code sees a single shape.

import Foundation

public enum KMLImportError: Error, CustomStringConvertible {
    case malformedXML(String)
    case noPolygons

    public var description: String {
        switch self {
        case .malformedXML(let r): return "Malformed KML: \(r)"
        case .noPolygons: return "KML document contains no Polygon features"
        }
    }
}

public enum KMLImporter {

    public static func importStrata(from data: Data) throws -> [ImportedPolygon] {
        let parser = KMLParser()
        let xml = XMLParser(data: data)
        xml.delegate = parser
        xml.shouldProcessNamespaces = true
        guard xml.parse() else {
            let desc = xml.parserError?.localizedDescription ?? "parse failed"
            throw KMLImportError.malformedXML(desc)
        }
        if let err = parser.error { throw err }
        guard !parser.polygons.isEmpty else { throw KMLImportError.noPolygons }
        return parser.polygons
    }
}

// MARK: - SAX parser

private final class KMLParser: NSObject, XMLParserDelegate {
    private enum RingKind { case outer, inner }

    var polygons: [ImportedPolygon] = []
    var error: Error?

    // Placemark-level state.
    private var placemarkName: String?
    private var placemarkAreaAcres: Double?
    // Polygon-level state (a Placemark may contain several via MultiGeometry).
    private var currentPolygonRings: [[CoordinateConversions.LatLon]] = []
    // Ring-level state.
    private var currentRingKind: RingKind?
    private var inCoordinates = false
    private var coordinateBuffer = ""
    private var inName = false
    private var nameBuffer = ""
    // Simple stack of element names so we know what context we're in.
    private var elementStack: [String] = []

    func parser(_ parser: XMLParser,
                didStartElement elementName: String,
                namespaceURI: String?,
                qualifiedName qName: String?,
                attributes attributeDict: [String: String] = [:]) {
        elementStack.append(elementName)
        switch elementName {
        case "Placemark":
            placemarkName = nil
            placemarkAreaAcres = nil
            currentPolygonRings = []
        case "Polygon":
            currentPolygonRings = []
        case "outerBoundaryIs":
            currentRingKind = .outer
        case "innerBoundaryIs":
            currentRingKind = .inner
        case "coordinates":
            inCoordinates = true
            coordinateBuffer = ""
        case "name":
            if elementStack.dropLast().last == "Placemark" {
                inName = true
                nameBuffer = ""
            }
        default:
            break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if inCoordinates { coordinateBuffer.append(string) }
        if inName { nameBuffer.append(string) }
    }

    func parser(_ parser: XMLParser,
                didEndElement elementName: String,
                namespaceURI: String?,
                qualifiedName qName: String?) {
        defer {
            if elementStack.last == elementName { elementStack.removeLast() }
        }
        switch elementName {
        case "coordinates":
            inCoordinates = false
            guard let kind = currentRingKind else { break }
            do {
                let ring = try parseCoordinateList(coordinateBuffer)
                switch kind {
                case .outer:
                    currentPolygonRings.insert(ring, at: 0)   // outer first
                case .inner:
                    currentPolygonRings.append(ring)
                }
            } catch let err {
                self.error = err
                parser.abortParsing()
            }
        case "outerBoundaryIs", "innerBoundaryIs":
            currentRingKind = nil
        case "name":
            if inName {
                inName = false
                placemarkName = nameBuffer.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        case "Polygon":
            flushPolygon()
        case "Placemark":
            // A Placemark containing a Polygon directly (no MultiGeometry) has
            // already been flushed. Reset state regardless.
            placemarkName = nil
            placemarkAreaAcres = nil
            currentPolygonRings = []
        default:
            break
        }
    }

    private func flushPolygon() {
        guard !currentPolygonRings.isEmpty, !currentPolygonRings[0].isEmpty else {
            currentPolygonRings = []
            return
        }
        let name = (placemarkName?.isEmpty == false) ? placemarkName! : "Stratum"
        let areaM2 = abs(GeoJSONImporter.signedPolygonAreaMetersSquared(rings: currentPolygonRings))
        let areaAcres = placemarkAreaAcres ?? GeoJSONImporter.metersSquaredToAcres(areaM2)
        let geoJSON = GeoJSONImporter.serialise(rings: currentPolygonRings)
        polygons.append(
            ImportedPolygon(
                name: name,
                areaAcres: areaAcres,
                rings: currentPolygonRings,
                geoJSONString: geoJSON
            )
        )
        currentPolygonRings = []
    }

    // MARK: - Coordinate list parsing

    /// Parse KML `coordinates` text: whitespace-separated `lon,lat[,alt]` tuples.
    private func parseCoordinateList(_ text: String) throws -> [CoordinateConversions.LatLon] {
        let tuples = text
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
        var ring: [CoordinateConversions.LatLon] = []
        ring.reserveCapacity(tuples.count)
        for tuple in tuples {
            let parts = tuple.split(separator: ",", omittingEmptySubsequences: true).map(String.init)
            guard parts.count >= 2,
                  let lon = Double(parts[0]),
                  let lat = Double(parts[1]) else {
                throw KMLImportError.malformedXML("Bad coordinate tuple: \(tuple)")
            }
            guard (-180...180).contains(lon), (-90...90).contains(lat) else {
                throw KMLImportError.malformedXML("lat/lon out of range in tuple: \(tuple)")
            }
            ring.append(CoordinateConversions.LatLon(latitude: lat, longitude: lon))
        }
        // KML does not require the ring to close; close it if needed.
        if let first = ring.first, let last = ring.last, first != last {
            ring.append(first)
        }
        if ring.count < 4 {
            throw KMLImportError.malformedXML("Ring has < 4 closed positions")
        }
        return ring
    }

    func parser(_ parser: XMLParser, parseErrorOccurred parseError: Error) {
        self.error = KMLImportError.malformedXML(parseError.localizedDescription)
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/geo/KMLImporter.kt
// KML stratum importer — 1:1 port of iOS Geo/KMLImporter.swift (spec §8,
// REQ-PRJ-002): import stratum boundaries from a Google-Earth-style KML
// document. Only `Placemark` elements carrying a `Polygon` (optionally nested
// under `MultiGeometry`) are extracted; other geometry types
// (Point/LineString/Model/etc.) are skipped.
//
// KML coordinates are expressed as `lon,lat[,alt]` whitespace-separated
// tuples. We reuse `ImportedPolygon` from `GeoJSONImporter` so downstream
// CruiseDesign code sees a single shape.
//
// Platform note: iOS uses the SAX-style XMLParser delegate; here the same
// state machine is driven from an XmlPullParser event loop (namespace-aware,
// so element names are local names — matching shouldProcessNamespaces).

package com.hcjeong.forestix.geo

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import org.xmlpull.v1.XmlPullParserFactory
import java.io.ByteArrayInputStream
import java.io.IOException

/// Import failure reasons; `description` strings match the iOS
/// `KMLImportError` cases byte-for-byte.
sealed class KMLImportError(val description: String) : Exception(description) {
    class MalformedXML(reason: String) : KMLImportError("Malformed KML: $reason")
    class NoPolygons : KMLImportError("KML document contains no Polygon features")
}

object KMLImporter {

    fun importStrata(data: ByteArray): List<ImportedPolygon> {
        val parser = KMLParser()
        try {
            val factory = XmlPullParserFactory.newInstance()
            factory.isNamespaceAware = true
            val xml = factory.newPullParser()
            xml.setInput(ByteArrayInputStream(data), null)
            var event = xml.eventType
            while (event != XmlPullParser.END_DOCUMENT) {
                when (event) {
                    XmlPullParser.START_TAG -> parser.didStartElement(xml.name)
                    XmlPullParser.TEXT -> parser.foundCharacters(xml.text ?: "")
                    XmlPullParser.END_TAG -> parser.didEndElement(xml.name)
                }
                event = xml.next()
            }
        } catch (e: KMLImportError) {
            throw e
        } catch (e: XmlPullParserException) {
            throw KMLImportError.MalformedXML(e.message ?: "parse failed")
        } catch (e: IOException) {
            throw KMLImportError.MalformedXML(e.message ?: "parse failed")
        }
        if (parser.polygons.isEmpty()) throw KMLImportError.NoPolygons()
        return parser.polygons
    }
}

// MARK: - SAX-style parser state machine

private class KMLParser {
    private enum class RingKind { OUTER, INNER }

    val polygons = mutableListOf<ImportedPolygon>()

    // Placemark-level state.
    private var placemarkName: String? = null
    private var placemarkAreaAcres: Double? = null

    // Polygon-level state (a Placemark may contain several via MultiGeometry).
    private var currentPolygonRings = mutableListOf<List<CoordinateConversions.LatLon>>()

    // Ring-level state.
    private var currentRingKind: RingKind? = null
    private var inCoordinates = false
    private var coordinateBuffer = StringBuilder()
    private var inName = false
    private var nameBuffer = StringBuilder()

    // Simple stack of element names so we know what context we're in.
    private val elementStack = mutableListOf<String>()

    fun didStartElement(elementName: String) {
        elementStack.add(elementName)
        when (elementName) {
            "Placemark" -> {
                placemarkName = null
                placemarkAreaAcres = null
                currentPolygonRings = mutableListOf()
            }
            "Polygon" -> currentPolygonRings = mutableListOf()
            "outerBoundaryIs" -> currentRingKind = RingKind.OUTER
            "innerBoundaryIs" -> currentRingKind = RingKind.INNER
            "coordinates" -> {
                inCoordinates = true
                coordinateBuffer = StringBuilder()
            }
            "name" -> {
                if (elementStack.getOrNull(elementStack.size - 2) == "Placemark") {
                    inName = true
                    nameBuffer = StringBuilder()
                }
            }
        }
    }

    fun foundCharacters(string: String) {
        if (inCoordinates) coordinateBuffer.append(string)
        if (inName) nameBuffer.append(string)
    }

    fun didEndElement(elementName: String) {
        try {
            when (elementName) {
                "coordinates" -> {
                    inCoordinates = false
                    val kind = currentRingKind
                    if (kind != null) {
                        val ring = parseCoordinateList(coordinateBuffer.toString())
                        when (kind) {
                            RingKind.OUTER -> currentPolygonRings.add(0, ring)   // outer first
                            RingKind.INNER -> currentPolygonRings.add(ring)
                        }
                    }
                }
                "outerBoundaryIs", "innerBoundaryIs" -> currentRingKind = null
                "name" -> {
                    if (inName) {
                        inName = false
                        placemarkName = nameBuffer.toString().trim()
                    }
                }
                "Polygon" -> flushPolygon()
                "Placemark" -> {
                    // A Placemark containing a Polygon directly (no MultiGeometry)
                    // has already been flushed. Reset state regardless.
                    placemarkName = null
                    placemarkAreaAcres = null
                    currentPolygonRings = mutableListOf()
                }
            }
        } finally {
            if (elementStack.lastOrNull() == elementName) {
                elementStack.removeAt(elementStack.size - 1)
            }
        }
    }

    private fun flushPolygon() {
        if (currentPolygonRings.isEmpty() || currentPolygonRings[0].isEmpty()) {
            currentPolygonRings = mutableListOf()
            return
        }
        val trimmedName = placemarkName
        val name = if (!trimmedName.isNullOrEmpty()) trimmedName else "Stratum"
        val areaM2 = kotlin.math.abs(
            GeoJSONImporter.signedPolygonAreaMetersSquared(currentPolygonRings)
        )
        val areaAcres = placemarkAreaAcres ?: GeoJSONImporter.metersSquaredToAcres(areaM2)
        val geoJSON = GeoJSONImporter.serialise(currentPolygonRings)
        polygons.add(
            ImportedPolygon(
                name = name,
                areaAcres = areaAcres,
                rings = currentPolygonRings,
                geoJSONString = geoJSON,
            )
        )
        currentPolygonRings = mutableListOf()
    }

    // MARK: - Coordinate list parsing

    /// Parse KML `coordinates` text: whitespace-separated `lon,lat[,alt]` tuples.
    private fun parseCoordinateList(text: String): List<CoordinateConversions.LatLon> {
        val tuples = text.split(Regex("\\s+")).filter { it.isNotEmpty() }
        val ring = ArrayList<CoordinateConversions.LatLon>(tuples.size)
        for (tuple in tuples) {
            val parts = tuple.split(',').filter { it.isNotEmpty() }
            if (parts.size < 2) {
                throw KMLImportError.MalformedXML("Bad coordinate tuple: $tuple")
            }
            val lon = parts[0].toDoubleOrNull()
            val lat = parts[1].toDoubleOrNull()
            if (lon == null || lat == null) {
                throw KMLImportError.MalformedXML("Bad coordinate tuple: $tuple")
            }
            if (lon !in -180.0..180.0 || lat !in -90.0..90.0) {
                throw KMLImportError.MalformedXML("lat/lon out of range in tuple: $tuple")
            }
            ring.add(CoordinateConversions.LatLon(latitude = lat, longitude = lon))
        }
        // KML does not require the ring to close; close it if needed.
        val first = ring.firstOrNull()
        val last = ring.lastOrNull()
        if (first != null && last != null && first != last) {
            ring.add(first)
        }
        if (ring.size < 4) {
            throw KMLImportError.MalformedXML("Ring has < 4 closed positions")
        }
        return ring
    }
}
